### DISCUSSION TITLE
custom client stuck waiting for server

### DISCUSSION THREAD

**Discussion by CFBerryhill** [iRobotEducation/create3_docs#597] [Questions] [answered]
### How are you connecting to your Create 3?

USB Ethernet (ROS 2)

### Computer(s) Model(s) and Operating System(s)

pi 4

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

H.2.3

### Which RMW is your robot running?

I don't know

### Is the robot connected to a network? If so, what is the network type?

pi is connected to a phone hotspot

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

None

### What is the Adapter Board's USB/BLE Toggle currently switched to?

None

### Describe your question.

Hi!

To elaborate on the setup - im sshing into a pi4 that is in the trunk of the create3 and connected via usb.

I want to send a simple 'drive to' command as detailed in [the documentation](https://iroboteducation.github.io/create3_docs/api/drive-goals/), which I can do no problem sending it directly through the command line!

`ros2 action send_goal /drive_distance irobot_create_msgs/action/DriveDistance "{distance: 0.5,max_translation_speed: 0.15}"`

Works just fine, as do a few other commands using different msgs. The issue is when I try to do the same thing with a python client. I followed [this tutorial](https://docs.ros.org/en/jazzy/Tutorials/Intermediate/Writing-an-Action-Server-Client/Py.html#writing-an-action-client) and changed the messages for the client and everything builds and compiles fine - but the robot doesnt move. Heres the code - copied from a "minimal" example from the linked tutorial.

```
import rclpy
from rclpy.action import ActionClient
from rclpy.node import Node
from irobot_create_msgs.action import DriveDistance

class MoveActionClient(Node):

    def __init__(self):
        super().__init__('move_action_client')
        self._action_client = ActionClient(self, DriveDistance, 'fibonacci')

    def send_goal(self, distance):
        goal_msg = DriveDistance.Goal()
        goal_msg.distance = distance

        print('waiting for server...')
        self._action_client.wait_for_server()

        print('sending goal...')
        return self._action_client.send_goal_async(goal_msg)

def main(args=None):
    rclpy.init(args=args)

    action_client = MoveActionClient()

    future = action_client.send_goal(0.5)

    rclpy.spin_until_future_complete(action_client, future)

if __name__ == '__main__':
    main()

```

And in the terminal all I see is `'waiting for server...'` and no movement. I have everything sourced and built with colcon, and I dont see why I can give the prompt directly but not through the code. Even if I run them  simultaneously from different terminals the code just can't find the server. But if i Ctrl+C and then immedietly run the above command it works after a slight lag - heres the output:
```

ubuntu@roombapi:~/create3-SLAM$ ros2 action send_goal /drive_distance irobot_create_msgs/action/DriveDistance "{distance: 0.5}"
Waiting for an action server to become available...
Sending goal:
     distance: 0.5
max_translation_speed: 0.3

Goal accepted with ID: d14e36a3097e40c798f57f50579c58e7

```

Very new to ROS2, so entirely possible I'm simply missing a step. Any help would be appreciated.

Note: heres my project structure. I'm building on top of some existing repos that i've simply added as submodules in the src directory. So it looks like:

- create3-SLAM
    - build
    - install
    - log
    - README.md
    - src
        - create3_examples
        - (some private packages that work independently)
        - move_robot_client.py (what I'm trying to run)

I think its possible that I've done something goofy here and thats what the issue is - I ran colcon build from 'foo' and it built with just some warnings about setup.py is deprecated. 

I think thats everything that has a snowballs chance in hell of being relevant, but happy to provide more details.

**Comment by alsora** [comment, marked as answer]
Hi @CFBerryhill,
this is the problem
```
self._action_client = ActionClient(self, DriveDistance, 'fibonacci')
```

You should indicate the name of the action service (which here is `drive_distance`, not `fibonacci` as in the example).
So the correct line would be
```
self._action_client = ActionClient(self, DriveDistance, 'drive_distance')
```

**Reply by CFBerryhill** [reply]
ohhhhhhh my god. thank you, knew it was going to be something silly. Would have realized that on my own eventually im sure, just didnt see it. One of those days...

Thanks!

**Reply by CFBerryhill** [reply]
aaaaand it immedietly worked. Thanks again.
